staging/vhost-device-gpu/src/virt_gpu.rs

```diff
@@ -453,6 +466,35 @@ impl RutabagaVirtioGpu {
             Err(_) => OkNoData,
         }
     }
+
+    fn read_2d_resource(
+        &mut self,
+        resource_id: u32,
+        rect: &Rectangle,
+        output: &mut [u8],
+    ) -> RutabagaResult<()> {
+        const BYTES_PER_PIXEL: usize = 4;
+        let result_len = rect.width as usize * rect.height as usize * BYTES_PER_PIXEL;
+        assert!(output.len() >= result_len);
+
+        let transfer = Transfer3D {
+            x: rect.x,
+            y: rect.y,
+            z: 0,
+            w: rect.width,
+            h: rect.height,
+            d: 1,
+            level: 0,
+            stride: rect.width * BYTES_PER_PIXEL as u32,
+            layer_stride: 0,
+            offset: 0,
+        };
+
+        self.rutabaga
+            .transfer_read(0, resource_id, transfer, Some(IoSliceMut::new(output)))?;
```

Comment:
shouldn't there be ctx_id here instead of 0?

Reply:
I don't know. ctx_id seems to be special when I looked at virglrenderer source code, but I don't know what is signifies. However crosvm seems to use 0 for this purpose too.
https://github.com/google/crosvm/blob/c1532a94f2895b003aa04df9f497a63082d4a3dc/devices/src/virtio/gpu/virtio_gpu.rs#L385-L395

Reply:
I also added a comment